contracts/src/Chainvoice.sol

```diff
@@ -1,104 +1,190 @@
-    mapping(address => uint256[]) public receivedInvoices; // Invoices received by a user
+    address public owner;
+    address public treasuryAddress;
+    uint256 public feeAmountInUSD;
 
-    event InvoiceCreated(uint256 id, address indexed from, address indexed to, uint256 amountDue);
-    event InvoicePaid(uint256 id, address indexed from, address indexed to, uint256 amountPaid);
+    constructor(address _priceFeed) {
+        priceFeed = AggregatorV3Interface(_priceFeed);
+        owner = msg.sender;
+        feeAmountInUSD = 1 * 1e18;
+    }
+
+    modifier OnlyOwner() {
+        require(msg.sender == owner, "Only Owner is accessible");
+        _;
+    }
+    event InvoiceCreated(
+        uint256 id,
+        address indexed from,
+        address indexed to,
+        uint256 amountDue
+    );
+    event InvoicePaid(
+        uint256 id,
+        address indexed from,
+        address indexed to,
+        uint256 amountPaid
+    );
 
-    /**
-     * @dev Create a new invoice request
-     * @param amountDue The amount requested in the invoice
-     * @param to The address of the receiver
-     */
-    function createInvoice(uint256 amountDue, address to) external {
-        console.log(to);
+    function createInvoice(
+        uint256 amountDue,
+        address to,
+        string memory _dueDate,
+        string memory _issueDate,
+        UserDetails memory user,
+        UserDetails memory client,
+        ItemData[] memory _items
+    ) external {
         require(amountDue > 0, "Amount must be greater than zero");
         require(to != address(0), "Receiver address cannot be zero");
         require(to != msg.sender, "Cannot create invoice for yourself");
 
-        // Create the invoice and push to the invoices array
         uint256 invoiceId = invoices.length;
-        invoices.push(InvoiceDetails({
-            id: invoiceId,
-            from: msg.sender,
-            to: to,
-            amountDue: amountDue,
-            isPaid: false
-        }));
-
-        // Track the invoice IDs for both sender and receiver
+        invoices.push(
+            InvoiceDetails({
+                id: invoiceId,
+                from: msg.sender,
+                dueDate:_dueDate,
+                issueDate: _issueDate,
+                user: user,
+                to: to,
+                client: client,
+                amountDue: amountDue,
+                isPaid: false
+            })
+        );
+
+        for (uint256 i = 0; i < _items.length; i++) {
+            itemDatas[invoiceId].push(_items[i]);
+        }
+
         sentInvoices[msg.sender].push(invoiceId);
         receivedInvoices[to].push(invoiceId);
 
         emit InvoiceCreated(invoiceId, msg.sender, to, amountDue);
     }
-    /**
-     * @dev Pay an invoice request using Ether
-     * @param invoiceId The ID of the invoice to be paid
-     */
+
+    function usdToNativeCurrencyConversion() public view returns (uint256) {
```

Comment:
Let's not do any conversion. Doing conversions using chainlink oracles would make us dependent on chainlink oracles.

Instead, let's allow the invoice to specify whether to pay in ETH or in any ERC20 token. Then, if an invoice is in USD, it is up to the users to choose and ERC20 token that it is a USD-pegged stablecoin.

Reply:
For this PR, let's focus on ETH invoices, though. Let's worry about ERC20 invoices in a future PR.